content/posts/2023-07-24-breakdown-of-http-clients-in-elixir/index.md

```diff
@@ -0,0 +1,458 @@
+
+```elixir
+defmodule ExampleApi do
+  use Tesla
+
+  plug Tesla.Middleware.BaseUrl, "http://api.example.com"
+  plug Tesla.Middleware.JSON
+  plug MyProject.CustomMiddleware
+
+  def fetch_users do
+    get("/users")
+  end
+end
+```
+
+There's some magic going on here, but I've used this API in the past in
+application code, and I find it pretty nice.
+
+The biggest advantage of Tesla in my opinion, however, is that it uses a
+"swappable" underlying HTTP adapter system. It supports [many built-in HTTP
+adapters][tesla-adapters], including ones for `httpc` and Finch.
+
+#### When to Use Tesla?
+
+I find Tesla and Req to be similar in level of abstraction and functionality, so
+I don't have a particular recommendation. I think Tesla, however, can be a good
+choice for library authors who need a rich HTTP client for their library's
+functionality. In those cases, Tesla makes a good candidate exactly because of
+its HTTP adapter system: as a library author, as long as you expose
+functionality to swap the Tesla adapter, you're not forcing your users to use a
+particular HTTP client. This is a big deal as far as I'm concerned, because I've
+worked on projects with tens and tens of dependencies. In those projects, many
+of those dependencies were a result of exactly this, that is, different
+libraries using different HTTP clients (and all their transitive dependencies).
+
+## Choosing the Right Client
+
+I've tried to write a bit about when to use each client so far, but to recap,
+this is my loose recommendation:
+
+| Client         | When                                                                                                                                                            |
+| :---------     | :-------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| *Mint*         | You need 100% control on connections and request lifecycle                                                                                                      |
+| *Mint*         | You already have a process architecture, and don't want to introduce any more processes                                                                         |
+| *Mint*         | You're a library author and you want to force as few dependencies as possible on your users                                                                     |
+| *Finch*        | You need a low-level client with high performance, but you want pooling built in                                                                                |
+| *Req*          | Scripting                                                                                                                                                       |
+| *Req*, *Tesla* | You want to build tailored HTTP clients in your application                                                                                                     |
```

Comment:
"Tailored" seems quite "advanced", I'm afraid people who are just looking for a lib to make some HTTP calls might not recognize themselves.